lib/modules/webserver/server.js

```diff
@@ -1,37 +1,63 @@
 let finalhandler = require('finalhandler');
+const async = require('async');
 let http = require('http');
 let serveStatic = require('serve-static');
 const {canonicalHost, defaultHost, dockerHostSwap} = require('../../utils/host');
 require('http-shutdown').extend();
 
 class Server {
   constructor(options) {
-    this.dist = options.dist || 'dist/';
+    this.buildDir = options.buildDir;
+    this.events = options.events;
     this.port = options.port || 8000;
     this.hostname = dockerHostSwap(options.host) || defaultHost;
+    this.isFirstStart = true;
+    this.opened = false;
   }
 
   start(callback) {
     if (this.server && this.server.listening) {
       let message = __("a webserver is already running at") + " " +
-                      ("http://" + canonicalHost(this.hostname) +
-                      ":" + this.port).bold.underline.green;
+          ("http://" + canonicalHost(this.hostname) +
+           ":" + this.port).bold.underline.green;
       return callback(null, message);
     }
-    let serve = serveStatic(this.dist, {'index': ['index.html', 'index.htm']});
+    let serve = serveStatic(this.buildDir, {'index': ['index.html', 'index.htm']});
 
     this.server = http.createServer(function onRequest(req, res) {
       serve(req, res, finalhandler(req, res));
     }).withShutdown();
 
+    const self = this;
 
-    this.server.listen(this.port, this.hostname, () => {
-      this.port = this.server.address().port;
-      callback(null, __("webserver available at") +
-        " " +
-        ("http://" + canonicalHost(this.hostname) +
-          ":" + this.port).bold.underline.green, this.port);
-    });
+    async.waterfall([
+      function createPlaceholderPage(next) {
+        if (self.isFirstStart) {
+          self.isFirstStart = false;
+          return self.events.request('build-placeholder', next);
+        }
+        next();
+      },
+      function listen(next) {
+        self.server.listen(self.port, self.hostname, () => {
+          self.port = self.server.address().port;
+          next();
+        });
+      },
+      function openBrowser(next) {
+        if (!self.opened) {
+          self.opened = true;
+          return self.events.request('open-browser', next);
+        }
+        next();
+      },
+      function reportAvailable(next) {
+        next(null, __("webserver available at") +
```

Comment:
I would personally put that one in the last callback, as it actually does nothing, but not a big deal.